### PR TITLE
ci: add static typing workflow

### LINKED ISSUE #47: Add `static-typing.yml` workflow file
See [here](https://github.com/ANNUBS/annubes/blob/main/.github/workflows/static-typing.yml) as an example.

### PR DESCRIPTION
**Pull request by gcroci2**
The configuration parameters will be moved to the .toml file in #51

### CODE REVIEW

**Comment by gcroci2** [comment]
Something weird is happening. The errors on the static typing that I see in the CI (a couple of errors in annubes.py) can't actually be found locally. But locally I find many more errors on other files (71 errors in 44 files). 

`mypy` version is the same (1.11.2)

Has this ever happened to you? Could you check what happens locally on your machine? @DaniBodor

**Comment by DaniBodor** [comment]
> Something weird is happening. The errors on the static typing that I see in the CI (a couple of errors in annubes.py) can't actually be found locally. But locally I find many more errors on other files (71 errors in 44 files).
> 
> `mypy` version is the same (1.11.2)
> 
> Has this ever happened to you? Could you check what happens locally on your machine? @DaniBodor

I also got "Found 71 errors in 44 files (checked 72 source files)". I am now recreating the env as it is defined on this branch and at this point in time and am curious to see what happens then.

**Comment by DaniBodor** [comment]
I re-created a conda env as defined in this branch, and still have way more issues than the workflow here:

> Found 71 errors in 44 files (checked 72 source files)

haven't checked the specific ones, to see whether these were found or not

**Comment by gcroci2** [comment]
I recreated an env using the toml file we have now (after having merged this branch with main) and everything now works identically to the CI. Now sure why before it wasn't working, but maybe the changes we introduced with the toml file did the job. Anyway, I'd advise you to recreate the env from scratch locally, using either this branch or the main. PR is now really ready :) @DaniBodor

**Comment by DaniBodor** [comment]
Creating a fresh env with python 3.11 from this branch (as of 75956af) still leads to "Found 71 errors in 44 files (checked 73 source files)" locally for me.
It seems to work on the remote and apparently on your machine, so that is odd. Not sure whether we just want to merge now or look further into that.

File: .github/workflows/static-typing.yml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+          which python3
+          python3 --version
+      - name: Upgrade pip and install dependencies
+        run: |
+          python3 -m pip install --upgrade pip setuptools
+          pip install mypy types-toml
+      # mypy configs are in pyproject.toml
+      - name: Check static typing
+        run: |
+          mypy neurogym tests
```
Comment (DaniBodor) [inline .github/workflows/static-typing.yml]: inserting final new line. It's odd that this didn't happen automatically for you, as it is set in the VS code settings files.

```suggestion
          mypy neurogym tests

```

File: mypy.ini [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Mypy: to check static type
+# see https://mypy.readthedocs.io/en/stable/config_file.html#example-pyproject-toml
+[mypy]
+python_version = 3.11
+warn_return_any = true
+warn_unused_configs = true
+ignore_missing_imports = true
+show_error_codes = true
```
Comment (DaniBodor) [inline mypy.ini]:
```suggestion
show_error_codes = true

```

File: neurogym/core.py [hunk 3]
```diff
@@ -69,7 +69,7 @@ def __init__(self, dt=100) -> None:
         self.t = self.t_ind = 0
         self.tmax = 10000  # maximum time steps
         self.performance = 0
-        self.rewards = {}
+        self.rewards: dict = {}
```
Comment (DaniBodor) [inline neurogym/core.py:72]: why do some need a type hint, and others not? Could this be the (or a) reason for failing mypy?

File: pyproject.toml [hunk 4]
```diff
@@ -60,7 +60,7 @@ dev = [
     "nbsphinx",
 ]
 publishing = ["build", "twine", "wheel"]
-tutorials = ["jupyter"]
+tutorials = ["notebook"]
```
Comment (DaniBodor) [inline pyproject.toml:61]: `build` is also part of the optional dev dependencies, but I don't think it needs to be there (just here for the publishing part).

File: neurogym/core.py (hunk 3 above)
Comment (gcroci2) [reply neurogym/core.py:72]: Usually type hints are inferred (that's why there are no hints in `self.tmax`, `self.performance`, etc.), but when initializing an empty dictionary mypy cannot infer what types the keys and values will have, so without the `dict` hint it gives an error.

File: pyproject.toml (hunk 4 above)
Comment (gcroci2) [reply pyproject.toml:61]: In the `build.yml` file we verify that we can build the package and we use `build`, that's why it's also in the `dev` list of deps

**Comment by gcroci2** [comment]
> Creating a fresh env with python 3.11 from this branch (as of [75956af](https://github.com/ANNUBS/neurogym/commit/75956afdd7512d133f33fe75d3980acd5ebe20c2)) still leads to "Found 71 errors in 44 files (checked 73 source files)" locally for me. It seems to work on the remote and apparently on your machine, so that is odd. Not sure whether we just want to merge now or look further into that.

I'd like to merge this anyway, indeed it is odd but I don't have time now to look further into it (also not sure how)

File: neurogym/core.py (hunk 3 above)
Comment (DaniBodor) [reply neurogym/core.py:72]: Then shouldn't the type hint include the types for the keys and values?

File: pyproject.toml (hunk 4 above)
Comment (DaniBodor) [reply pyproject.toml:61]: we actually use both dependency lists in build.yml anyway. Not a huge issue, just seems odd to state the same package twice:

https://github.com/ANNUBS/neurogym/blob/83792a3634ef9428828afd88e8fa8a2a4890f295/.github/workflows/build.yml#L54

**Review by DaniBodor — APPROVED** [review]

File: pyproject.toml (hunk 4 above)
Comment (gcroci2) [reply pyproject.toml:61]: oh yes you're right, I've just removed `build` from the `dev` list

File: neurogym/core.py (hunk 3 above)
Comment (gcroci2) [reply neurogym/core.py:72]: it can but not necessary, with the dict only the mypy issue gets fixed

**Comment by DaniBodor** [comment]
why not add RUFF012 to the ignore list rather than ignoring it so many times?

**Comment by gcroci2** [comment]
> why not add RUFF012 to the ignore list rather than ignoring it so many times?

I don't think it's good to ignore that rule, for sure we don't want to ignore in cases we create new classes attributes, and even in the already present cases we should verify what was the purpose and why the attributes are being overridden that way. If at a certain point we realize that we will never look into it, we can always place the rule in the ignore list later.

**Comment by gcroci2** [comment]
> > why not add RUFF012 to the ignore list rather than ignoring it so many times?
> 
> I don't think it's good to ignore that rule, for sure we don't want to ignore in cases we create new classes attributes, and even in the already present cases we should verify what was the purpose and why the attributes are being overridden that way. If at a certain point we realize that we will never look into it, we can always place the rule in the ignore list later.

**Pull request merged**